sklearn/model_selection/_search.py

```diff
@@ -819,30 +880,45 @@ def evaluate_candidates(candidate_params, cv=None, more_results=None):
                         )
                     )
 
-                out = parallel(
-                    delayed(_fit_and_score)(
-                        clone(base_estimator),
-                        X,
-                        y,
-                        train=train,
-                        test=test,
-                        parameters=parameters,
-                        split_progress=(split_idx, n_splits),
-                        candidate_progress=(cand_idx, n_candidates),
-                        **fit_and_score_kwargs,
-                    )
-                    for (cand_idx, parameters), (split_idx, (train, test)) in product(
-                        enumerate(candidate_params), enumerate(cv.split(X, y, groups))
+                def _generate_jobs():
```

Comment:
With this much indentation, I think we can have `_generate_jobs` be a helper function. Something like this:

```python
def _generate_jobs(
    *,
    splits,
    base_estimator,
    use_warm_start,
    candidate_params,
    n_candidates,
    fit_and_score_kwargs,
):
    cand_idx = 0
    n_splits = len(splits)
    warm_start_groups = _generate_warm_start_groups(candidate_params, use_warm_start)

    for warm_candidates in warm_start_groups:
        for (split_idx, (train, test)) in enumerate(splits):
            yield delayed(_warm_fit_and_score)(
                clone(base_estimator),
                warm_candidates,
                train=train,
                test=test,
                split_progress=(split_idx, n_splits),
                cand_idx=cand_idx,
                n_candidates=n_candidates,
                **fit_and_score_kwargs,
            )

        cand_idx += len(warm_candidates)
```

And then call it in `evaluate_candidates`:

```python
out = parallel(
    _generate_jobs(
        splits=list(cv.split(X, y, groups)),
        base_estimator=base_estimator,
        use_warm_start=getattr(self, "use_warm_start", None),
        candidate_params=candidate_params,
        n_candidates=n_candidates,
        fit_and_score_kwargs=fit_and_score_kwargs,
    )
)

```

This would also workaround the numpydoc + yield issue.

Reply:
yes, I've considered pulling it out, not had the chance. Thanks for doing some of the work for me!

Reply:
The problem is managing the closure context, but maybe you've sorted that out